test/distributed/org/apache/cassandra/distributed/upgrade/MixedModeBatchTestBase.java

```diff
@@ -0,0 +1,110 @@
+import org.apache.cassandra.exceptions.WriteFailureException;
+
+import static org.apache.cassandra.distributed.shared.AssertUtils.assertRows;
+import static org.apache.cassandra.distributed.shared.AssertUtils.row;
+
+public class MixedModeBatchTestBase extends MixedModeReplicationTestBase
+{
+    private static final int KEYS_PER_BATCH = 10;
+
+    protected void testSimpleStrategy(Versions.Major from, Versions.Major to, boolean isLogged) throws Throwable
+    {
+        String insert = "INSERT INTO test_simple.names (key, name) VALUES (%d, '%s')";
+        String select = "SELECT * FROM test_simple.names WHERE key = ?";
+
+        new TestCase()
+        .nodes(3)
+        .nodesToUpgrade(1, 2)
+        .upgrade(from, to)
+        .setup(cluster -> {
+            cluster.schemaChange("CREATE KEYSPACE test_simple WITH replication = {'class': 'SimpleStrategy', 'replication_factor': 2};");
+            cluster.schemaChange("CREATE TABLE test_simple.names (key int PRIMARY KEY, name text)");
+        })
+        .runAfterNodeUpgrade((cluster, upgraded) -> {
+            List<Long> initialTokens = new ArrayList<>(cluster.size() + 1);
+            initialTokens.add(null); // The first valid token is at 1 to avoid offset math below.
+
+            for (int i = 1; i <= cluster.size(); i++)
+                initialTokens.add(Long.valueOf(cluster.get(i).config().get("initial_token").toString()));
+
+            List<Long> validTokens = initialTokens.subList(1, cluster.size() + 1);
+
+            // Exercise all the coordinators...
+            for (int i = 1; i <= cluster.size(); i++)
+            {
+                StringBuilder batchBuilder = new StringBuilder("BEGIN " + (isLogged ? "" : "UNLOGGED ") + "BATCH\n");
+                String name = "Julia";
+                Runnable[] tests = new Runnable[KEYS_PER_BATCH];
+
+                // ...and sample enough keys that we cover the ring.
+                for (int j = 0; j < KEYS_PER_BATCH; j++)
+                {
+                    int key = j + (i * KEYS_PER_BATCH);
+                    batchBuilder.append(String.format(insert, key, name)).append('\n');
+
+                    // Track the test that will later verify that this mutation was replicated properly.
+                    tests[j] = () -> {
+                        Object[] row = row(key, name);
+                        Long token = tokenFrom(key);
+                        int node = calculatePrimary(validTokens, token);
+                        assertRows(cluster.get(node).executeInternal(select, key), row);
+
+                        node = nextNode(node, cluster.size());
+                        assertRows(cluster.get(node).executeInternal(select, key), row);
+
+                        // At RF=2, this node should not have received the write.
+                        node = nextNode(node, cluster.size());
+                        assertRows(cluster.get(node).executeInternal(select, key));
```

Comment:
TODO: These assertions probably need better supporting information on failure, similar to what happens below for writes.

Reply:
Going to resolve this. The trace is actually pretty good if one of these assertions fails.